File: big-o-notation/exercises/is_palindrome/is_palindrome.go
package main

import (
	"fmt"
	"strings"
)

// IsPalindrome checks if a string is a palindrome using the two-pointer approach.
// Time Complexity: O(n) - where n is the length of the string.
// Space Complexity: O(1) - as we are only using two pointers.
func IsPalindrome(s string) bool {
	// Normalize the string (lowercase for comparison)
	s = strings.ToLower(s)
	runes := []rune(s)

	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		if runes[i] != runes[j] {
			return false
		}
	}
	return true
}

func main() {
	testCases := []string{"Racecar", "hello", "A man a plan a canal Panama"}

	for _, tc := range testCases {
		// Note: The simple version above considers spaces.
		// For "A man a plan a canal Panama" to be true, we'd need to strip spaces.
		// For this basic exercise, we'll keep it simple.
		fmt.Printf("Is '%s' a palindrome? %v\n", tc, IsPalindrome(tc))
	}
}

File: big-o-notation/exercises/is_palindrome/is_palindrome.rs
fn is_palindrome(s: &str) -> bool {
    // Time Complexity: O(n)
    // Space Complexity: O(n) (for the reversed string)
    let normalized = s.to_lowercase().replace(|c: char| !c.is_alphanumeric(), "");
    let reversed: String = normalized.chars().rev().collect();
    normalized == reversed
}

fn is_palindrome_optimized(s: &str) -> bool {
    // Time Complexity: O(n)
    // Space Complexity: O(1) (ignoring the temporary lowercase string if needed)
    let normalized = s.to_lowercase();
    let chars: Vec<char> = normalized.chars().filter(|c| c.is_alphanumeric()).collect();

    if chars.is_empty() {
        return true;
    }

    let mut left = 0;
    let mut right = chars.len() - 1;

    while left < right {
        if chars[left] != chars[right] {
            return false;
        }
        left += 1;
        right -= 1;
    }
    true
}

fn main() {
    let test_cases = vec!["Racecar", "hello", "A man, a plan, a canal: Panama"];
    for tc in test_cases {
        println!(
            "Is \"{}\" a palindrome? {}",
            tc,
            is_palindrome_optimized(tc)
        );
    }
}
